src/java/org/apache/cassandra/security/DefaultSslContextFactory.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.security;
+
+import java.io.File;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.KeyStore;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.TrustManagerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Cassandra's default implementation class for the configuration key {@code ssl_context_factory}. It uses
+ * file based keystores.
+ */
+public final class DefaultSslContextFactory extends AbstractSslContextFactory
```

Comment:
Thanks - this is along the lines of what I was thinking.  I think having the `DefaultSslContextFactory` as final is a good thing but it would prevent the use case of somebody who just wants to just do minor tweaks the the netty SSL context and keep all of the netty machinery in place.

What do you think about renaming this to a non-final `HotReloadingSslContextFactory` class and making a shell `final class DefaultSslContextFactory extends HotReloadingSslContextFactory`?

Either that or we drop the `final` from this class so it can be extended, but not sure if that opens up any security issues - probably not as an attacker would have had sufficient credentials to alter the config to pick up a hostile parameterized class anyway.

Reply:
I created FileBasedSslContextFactory which is abstract and has File Based key/trust stores support + Hot-reloading assuming the file based system. I created the the DefaultSslContextFactory as a shell class. I was thinking about PEM format files instead of standard keystore formats and this model would fit in better with that.